service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java

```diff
@@ -506,6 +506,35 @@ protected void doStop(final Future<Void> stopFuture) {
                 });
     }
 
+    /**
+     * Checks if this adapter may accept another telemetry or event message from a device.
+     * <p>
+     * This default implementation uses the
+     * {@link ResourceLimitChecks#isMessageLimitReached(TenantObject tenant, long payloadSize)} method
+     * to verify if the tenant's message limit has been reached.
+     *
+     * @param tenantConfig The tenant to check the message limit for.
+     * @param payloadSize  The size of the message payload in bytes.
+     * @return A succeeded future if the message limit has not been reached yet
+     *         or if the limits could not be checked.
+     *         Otherwise the future will be failed with a {@link ClientErrorException}
+     *         containing the 403 Forbidden status code.
+     * @throws NullPointerException if tenant is {@code null}.
+     */
+    protected Future<Void> checkMessageLimit(final TenantObject tenantConfig, final long payloadSize) {
+
+        Objects.requireNonNull(tenantConfig);
+        return resourceLimitChecks.isMessageLimitReached(tenantConfig, payloadSize)
+                .recover(t -> Future.succeededFuture(Boolean.FALSE))
+                .compose(isExceeded -> {
+                    if (isExceeded) {
+                        return Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_FORBIDDEN));
```

Comment:
Wouldn't it make sense to reply with `429 Too Many Requests`?

https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429

Reply:
`429 Too Many Requests` suits here quite well.